src/main/java/org/candlepin/async/JobMessageReceiver.java

```diff
@@ -142,30 +163,42 @@ private CPMSession createSession() throws CPMException {
     /**
      * Starts all of the managed sessions. If a given session has died or been closed, this method
      * will recreate it.
+     *
+     * @throws CPMException
+     *  if there is a session that cannot be recreated or started
      */
     private void startSessions() throws CPMException {
-        Set<CPMSession> created = null;
-
-        Iterator<CPMSession> iterator = this.sessions.iterator();
-        while (iterator.hasNext()) {
-            CPMSession session = iterator.next();
-
-            if (session == null || session.isClosed()) {
-                if (created == null) {
-                    created = new HashSet<>();
+        Set<CPMSession> updatedSessions = new HashSet<>();
+        CPMException bufferedException = null;
+        int openSessionCount = 0;
+        for (CPMSession session : this.sessions) {
+            try {
+                if (bufferedException != null) {
+                    continue;
                 }
 
-                iterator.remove();
-                session = this.createSession();
+                if (session == null || session.isClosed()) {
+                    session = this.createSession();
+                }
 
-                created.add(session);
+                session.start();
+            }
+            catch (CPMException exception) {
+                bufferedException = exception;
+            }
+            finally {
+                updatedSessions.add(session);
+                if (!session.isClosed()) {
```

Comment:
Doesn't this throw an NPE if the session isn't created. Like when it goes after the catch block?

Reply:
Good catch! Updated